### ISSUE TITLE
runtime.sendMessage/onMessage

### ISSUE THREAD

**Issue by hgwood** [mozilla/webextension-polyfill#58]
Hello

I'm running into problems when trying to use `browser.runtime.sendMessage` and `browser.runtime.onMessage` with this polyfill on Chrome 61.

I have this code that works fine:

```js
// in background script, run when the extension is loaded
chrome.runtime.onMessage.addListener((message, sender) => {
  console.log("background script got message", message, "from", sender);
});

// in content script, run when the user presses a button
chrome.runtime.sendMessage({type: "event", eventName: "THE_EVENT"});
```

If I replace `chrome` with `browser` in the background script, I get this in its console:

![image](https://user-images.githubusercontent.com/1656170/30859797-63b67cf4-a2c5-11e7-97b1-a4c2930b27d6.png)

and the listener is *not* run.

Independently, if I replace `chrome` with `browser` on the content script side, I get this in the console of the page:

![image](https://user-images.githubusercontent.com/1656170/30859893-ab292d48-a2c5-11e7-9979-c781970f8639.png)

but the listener *is* run.

If I replace `chrome` with `browser` on both sides, then the content script says this instead:

![image](https://user-images.githubusercontent.com/1656170/30859985-01e267d0-a2c6-11e7-9f6b-7cf2cb501f55.png)

and the background script message remains unchanged. 

I'm new to writing browser extensions so I'm not sure if this might be a bug in the polyfill or if not using the API quite right.

**Comment by hgwood** [comment]
OK I think the two errors always happen but they are not reported when using `chrome` because they are stored in `runtime.lastError`, while the polyfill uses rejected promises, which do appear as console errors. So only the background script side remains a problem.

**Comment by Rob--W** [comment]
(tested with Chrome 61 and latest polyfill version (0.1.2))

If I use `chrome` at the content script and `browser` at the background script, the message gets through without errors.
If I use `browser` at the content script and `chrome` or `browser` at the background script, the message also gets through, but the error message is always set:

> The message port closed before a response was received.

I think that we should catch this specific error and ignore it (treat it as no response).

**Comment by jowa0720** [comment]
For me, it does not matter which combination of `browser` or `chrome` I use. In all cases the onMessage listener set in the background script is not executed.

If I use `browser` in the the background script, chrome console ouputs the following:

![error1](https://user-images.githubusercontent.com/14321098/31082257-bb352dd6-a78e-11e7-9531-721483c22f12.PNG)

If I use `chrome` in the the background script, the console adds an additional error:

![error2](https://user-images.githubusercontent.com/14321098/31082308-ddb11e7e-a78e-11e7-9f19-536639a2da54.PNG)

**Comment by Rob--W** [comment]
@rpl Could you investigate? I think that it might be related to #57.

**Comment by ahmadassaf** [comment]
i can confirm that #57 fixes the issues mentioned above in a sense that `browser` both in content and background script will be able to communicate. However, the issue now is the `sendResponse` in case needed to send back an answer does not work which i believe is somehow related to #16

**Comment by hgwood** [comment]
@ahmadassaf @Rob--W I was already using what is now named 0.2.0 when I witnessed and opened the issue, as I built the polyfill from `master`, so #57 was already included. I do not use `sendResponse`.

**Comment by hgwood** [comment]
It appears Chrome 61 has been updated since I opened the issue. Without changing my code, here is what I now get in the content script console when I use `browser` on the content script side:

![image](https://user-images.githubusercontent.com/1656170/31127348-eedaf114-a84e-11e7-8ef9-c09d3acc4896.png)

Looks similar to @FayneAldan and @jowa0720, but slightly different. Might be because of slightly different versions of Chrome. My version right now is 61.0.3163.100.

Anyway, this error does not stop the message from going through. However, the message does *not* get through when I use `browser` on the background script side. Polyfill version is still 0.2.0.

**Comment by ahmadassaf** [comment]
@hgwood i have just noticed something in my `background.js` that might be useful. Basically before coming across #57 i had the following:

```
const messageListner = typeof InstallTrigger !== 'undefined' ?  browser.runtime : chrome.extension;
```

Since i want my extension to work mainly in Chrome and FF, i added this check for FF and then assigning the message listener accordingly. I do no use `messageListner` anywhere in my codebase now .. but removing it make me get the error again .. just though to share this if it helps

**Comment by FayneAldan** [comment]
So, that would mean there's something wrong in the [Proxy](https://github.com/mozilla/webextension-polyfill/blob/318271214f55a0c4bb2f97b986ac501fc3541881/src/browser-polyfill.js#L274)'s [get](https://github.com/mozilla/webextension-polyfill/blob/318271214f55a0c4bb2f97b986ac501fc3541881/src/browser-polyfill.js#L200) method, right? Why do we even need proxies? Why not just make an object with the available methods?

**Comment by midzelis** [comment]
Just want to chime in and say that I'm currently hitting this issue. I'm going to see if browser specific workaround posted by ahmadassaf will work for me.

**Comment by DamienCassou** [comment]
My code
- works perfectly with webextension-polyfill 0.1.2
- fails with "Previous API instantiation failed." and "Could not establish connection. Receiving end does not exist." with version 0.2.0 and Chromium 60

The popup does:

```javascript
  await browser.tabs.executeScript(tabId, {file: '/browser-polyfill.min.js'})
  await browser.tabs.executeScript(tabId, {file: '/content-script.js'})
  await browser.tabs.sendMessage(tab.id, {action: 'getHeadings'})
```

and `content-script.js` is:

```javascript
function messageReceived (message) {
  return Promise.resolve(...)
}

browser.runtime.onMessage.addListener(messageReceived)
```

I suggest using 0.1.2 for now.

**Comment by hgwood** [comment]
I confirm that 0.1.2 also works for me.

**Comment by midzelis** [comment]
My workaround was to create my own polyfill, using https://github.com/tfoxy/chrome-promise as a start, and filling in the non-callback ones with a simple re-bind from browser to chrome.

**Comment by rpl** [comment]
The fix for this issues (#64) has been included in the 0.2.1 release (https://github.com/mozilla/webextension-polyfill/releases/tag/0.2.1) and published on npm.

**Comment by ElForastero** [comment]
I'm not sure what went wrong, but I get this error too.
v0.4.0.

Both background and content scripts use `browser.runtime` to send and receive messages.

![image](https://user-images.githubusercontent.com/5102818/65150572-760a0200-da35-11e9-84b7-4356d9d11646.png)
